### ISSUE TITLE
No equalizer in Realtek Audio Console

### ISSUE THREAD

**Issue by wkonowrocki** [pal1000/Realtek-UAD-generic#36]
I have native UAD driver on my custom build PC (asrock motherboard/audio) on that PC in Realtek Audio Console under Speakers I have an equalizer. I got myself a laptop (Lenovo Ideapad 300-15IKB) and installed UAD driver through your utility. Everything works except that in Realtek Audio Console I don't have an equalizer, only "Omni Speakers" switch.

**Comment by pal1000** [comment]
Maybe it was disabled or is unsupported by extension component as I was unable to update it since 6.0.8825,1.

Search your [sound card hardware ID](http://www.devicedriverfinder.com/blog/device-drivers/how-to-check-the-hardware-id-for-a-windows-device) [here](https://github.com/alanfox2000/realtek-universal-audio-driver).
Make sure you reduce it to one of these formats before searching;
- `HDAUDIO\FUNC_01&VEN_10EC&DEV_XXXX&SUBSYS_XXXXXXXX`
- `INTELAUDIO\FUNC_01&VEN_10EC&DEV_XXXX&SUBSYS_XXXXXXXX`

Please reply with link to search results. If your hardware ID is found then your laptop has Realtek UAD support. If it does you may try the OEM specific driver. Hopefully that has equalizer assuming your laptop is eligible.

**Comment by wkonowrocki** [comment]
I've done the steps you've provided and I was able to found a laptop with same sound card hardware ID - [ext_lenovo_330-15icn_rs4_rs5_19h1_v223](https://github.com/alanfox2000/realtek-universal-audio-driver/blob/master/ThirdParty/Dolby/20200211_dolbyapo_realtek/ext_realtek_lenovo_ideapad/ext_lenovo_330-15icn_rs4_rs5_19h1_v223/HDAUDIO_DEV_0230_SUBSYS_17AA3820_PCI_SUBSYS_383717AA.xml). This is one of many that popped up when I searched my sound card hardware ID, but it was the closest one that also matched my name. BTW my sound card hardware ID is `HDAUDIO_DEV_0230_SUBSYS_17AA3820`

**Comment by pal1000** [comment]
Found your hardware ID [with this search](https://github.com/alanfox2000/realtek-universal-audio-driver/search?q=HDAUDIO%5CFUNC_01%26VEN_10EC%26DEV_0230%26SUBSYS_17AA3820&unscoped_q=HDAUDIO%5CFUNC_01%26VEN_10EC%26DEV_0230%26SUBSYS_17AA3820). By looking at the results it looks like it uses both Dolby and Fortemedia extensions. It's no wonder equalizer doesn't work. It probably only works with Dolby APO.

Getting it to work won't be easy and my expertise at crafting or finding OEM specific UAD packages is very limited. I recommend posting one of these forums:
- [station-drivers](https://www.station-drivers.com/index.php?option=com_kunena&view=topic&catid=18&id=17&Itemid=858&lang=en );
- [My Digital Life](https://forums.mydigitallife.net/threads/update-realtek-high-definition-audio.72236/);
- [Tenforums](https://www.tenforums.com/sound-audio/135259-latest-realtek-hd-audio-driver-version-2-a.html)

Make sure you post hardware ID or link to search I posted.

**Comment by epman100** [comment]
I can pinpoint the wkonowrocki's problem with "no equalizer" and it lies within the extension INF file - HDX_LenovoExt_DOLBY_FORTE.inf.  not a problem with the unofficial generic uad drivers but with using certain "extension INF" files that came originally with the Lenovo based UAD drivers for his Lenovo computer.

that file seems to load the Dolby and Forte apo extensions & audio effects and use those exclusively and the Realtek APO audio effects options like the "equalizer" are not available (cuz Lenovo probably told Realtek to not use those Realtek EQ features while Dolby & Forte options are being used)

Compare the entries of the "APO CLSID registration for render endpoints" lines from the **HDX_LenovoExt_DOLBY_FORTE.inf** file and with another OEM/Dolby extension INF file like **HDX_AcerExt_DOLBY_RTK.inf** file.  The "render endpoints" section of the HDX_AcerExt_DOLBY_RTK.inf file has some additional lines (one mentioning %REALTEK_OMFX_CLSID%), which seem to allow usage of the Realtek audio effect options like equalizer while also using Dolby effects on certain Acer machines (maybe because Acer wanted to integrate Dolby while preserving or maintaining the Realtek audio options).

I may edit this later on if I have more info

**Comment by pal1000** [comment]
So you are suggesting he didn't follow [these instructions] or failed somewhere. A drivers list export from DriverStore Explorer can shed some light on this.

**Comment by Hazbelll** [comment]
My case is similar to @wkonowrocki's as I too have a Lenovo manufactured laptop (Realtek Codec's Hardware ID: `HDAUDIO\FUNC_01&VEN_10EC&DEV_0298&SUBSYS_17AA224D`) , but also lack the Realtek Equaliser in the Realtek Audio Console UWP App, with only the option to toggle "Omni Speaker" for the built-in speakers, and nothing for the external headphone jack.

As @epman100 leads on to, it's because these machines normally use Realtek legacy HDA drivers which uses the Dolby APO Extensions as a replacement for the Realtek Equaliser for the Speakers/Playback/Outputs and Forte APO Extensions for additional enhancement options available to the Microphone/Recording/Inputs.

Forte's APO Extensions are just additional enhancements added into the native list of available enhancements found in the legacy Realtek HD Audio Manager Win32 Application, or under the "Enhancements" tab of an Audio Endpoint (Input in this instance).

Annoyingly, Dolby's APO Extensions are controlled through a 3rd party program; In Lenovo's case in their [Lenovo Vantage UWP Application](https://www.microsoft.com/en-gb/p/lenovo-vantage/9wzdncrfj4mv). It is not at all present in the Realtek HD Audio Manager Win32 Application, nor is it included under the "Enhancements" tab of any Audio Endpoint (Outputs) has no options, other than a checkbox to universally disable enhancements (A separate tab exists instead for Dolby, but only with 1 button to globally enable/disable it).

From my own usage, the loss of Forte's APO Extensions is of no real loss. Realtek's own Noise Suppression and AEC (Acoustic Echo Cancellation) effects are available and work via Realtek's Audio Console UWP App, but not having the Realtek Equaliser is a bit of a bummer. The sound is by default, flat (No Equaliser), but honestly that's much better than having some terrible Dolby preset applied without any option to change/disable it.

**Comment by epman100** [comment]
> 
> 
> So you are suggesting he didn't follow [these instructions](https://github.com/pal1000/Realtek-UAD-generic#switching-from-other-realtek-hd-audio-drivers) or failed somewhere. A drivers list export from DriverStore Explorer can shed some light on this.

until wkonowrocki can post back here with the Driverstore Explorer screenshot we won't know

most likely the latter - failed somewhere
he either has to use a modified HDX_LenovoExt_DOLBY_FORTE.inf extension file with the added "OMFX" entries or clean install any unofficial generic uad using HDX_GenericExt_RTK.inf but not install any 3rd party SWC components like Dolby and Forte - dax3_ext_dolbyaudio.inf needs to be removed using Driverstore Explorer if that extension inf file exists.

by the way, I checked the Lenovo support site for a "Ideapad 300-15IKB" and that specific model does **not** exist (produced no results).  Maybe he meant "Ideapad **330**-15IKB", which the 330 model does exist.

**Comment by pal1000** [comment]
Not necessarily, @epman100, considering last paragraph from @Hazbelll's comment.

**Comment by Hazbelll** [comment]
Perhaps I should also mention in my case, I am using Driver Version: [6.0.8856.1](https://github.com/pal1000/Realtek-UAD-generic/releases/tag/6.0.8856.1).

My reasoning for this is I prefer using the Official Installer and not too fussed about having the very latest Driver version (From past experience, they pretty much add no benefit to functionality/stability unless between major OS revisions). This Driver Version is the latest that is fully WHQL Compliant I believe.

I don't believe this should make any difference however? On a past Windows installation, I used your Installer and the result was the same - No Realtek Equaliser and no OEM features.

**Comment by Cryosim** [comment]
Hi. Same here. missing equalizer for speakers and headphones. what happened?
I install your last drivers and replace in device drivers.
My sound ID
HDAUDIO\FUNC_01&VEN_10EC&DEV_1168&SUBSYS_104387C5&REV_1001
HDAUDIO\FUNC_01&VEN_10EC&DEV_1168&SUBSYS_104387C5

**Comment by pal1000** [comment]
@Cryosim, have you followed [this guide](https://github.com/pal1000/Realtek-UAD-generic#switching-from-other-realtek-hd-audio-drivers) to migrate otherwise you'll have a lot of problems including lack of  equalizer.

It looks like your sound card is UAD supported and uses [Asus Avolute APO with XPERI extension](https://github.com/alanfox2000/realtek-universal-audio-driver/search?q=HDAUDIO%5CFUNC_01%26VEN_10EC%26DEV_1168%26SUBSYS_104387C5&unscoped_q=HDAUDIO%5CFUNC_01%26VEN_10EC%26DEV_1168%26SUBSYS_104387C5).
Maybe it's disabled like for others. [This guide](https://github.com/alanfox2000/realtek-universal-audio-driver/wiki) from @alanfox2000 may help with creation of an OEM specific UAD package.

**Comment by Cryosim** [comment]
@pal1000 Yes. is onboard sound on motherboard Asus rog strix x570 E gaming.

**Comment by Cryosim** [comment]
@pal1000 So, disable internet, and uninstall drivers Programs and Features. stop service Windows audio. Reboot PC, and install again your drivers. Where i can find Audio Console?

**Comment by pal1000** [comment]
Audio Console is fetched via Windows store. You can wait or do a check for updates in Windows store.
Considering how much time passed you probably already have it @Cryosim

**Comment by Brainiarc7** [comment]
A similar case here on an EVOC P775TM1-R.
The hardware ID is `HDAUDIO\FUNC_01&VEN_10EC&DEV_0899&SUBSYS_15587711`, and the query returns a result [here](https://github.com/alanfox2000/realtek-universal-audio-driver/search?q=HDAUDIO%5CFUNC_01%26VEN_10EC%26DEV_0230%26SUBSYS_17AA3820&unscoped_q=HDAUDIO%5CFUNC_01%26VEN_10EC%26DEV_0230%26SUBSYS_17AA3820).

Main page, showing details:

![image](https://user-images.githubusercontent.com/2356871/75606582-36c10080-5aff-11ea-9689-c9ac5343057b.png)

No equalizer shown:

![image](https://user-images.githubusercontent.com/2356871/75606592-4dffee00-5aff-11ea-815b-e13c159c668b.png)

**Comment by pal1000** [comment]
@Brainiarc7 it looks like your sound card uses Dolby APO but without Fortemedia extension. I think OEM package is needed for equalizer, see [compatibility note](https://github.com/pal1000/Realtek-UAD-generic#compatibility).

**Comment by epman100** [comment]
actually HWID "HDAUDIO\FUNC_01&VEN_10EC&DEV_0899&SUBSYS_15587711" should produce **[this result](url=https://github.com/alanfox2000/realtek-universal-audio-driver/search?q=HDAUDIO\FUNC_01%26VEN_10EC%26DEV_0899%26SUBSYS_15587711&unscoped_q=HDAUDIO\FUNC_01%26VEN_10EC%26DEV_0899%26SUBSYS_15587711)**. which uses Creative APO extension rather than Dolby APO

**Comment by Hazbelll** [comment]
Interesting, thanks for the information @epman100.

This has sparked my curiosity however as I wonder if it's possible to enable the Realtek Audio Console's EQ and Environment Options for these generic UAD Drivers, ignoring whatever the respective OEM desired, without breaking WHQL Compliancy in some way.

I wouldn't know personally, I'm not that knowledgeable when it comes to modding/hacking drivers together like that, but it would be incredible if it were to be possible and would make this a truly generic UAD Driver.

**Comment by pal1000** [comment]
@Hazbelll, force updater has some patching ability. It can apply registry changes from HKR commands from INF files. Unfortunately it doesn't support extension patching. It can be added though.

**Comment by Hazbelll** [comment]
@pal1000 Would that be something you would be interested in adding? Or is it more complicated than that?

**Comment by pal1000** [comment]
I'll be working on this, but it will it take a while.

**Comment by epman100** [comment]
result of using that modded HDX_GigabyteExt_Creative_RTK.inf file

![realtek-uwp-creative-uwp-soundoptions-sbx720](https://user-images.githubusercontent.com/61091896/76191925-b012c000-619d-11ea-9c71-dc30b3834897.png)

oh yeah! :D

**Comment by Hazbelll** [comment]
> I'll be working on this, but it will it take a while.

That's great to hear @pal1000! Hope it goes smoothly for you, keep us posted.

**Comment by epman100** [comment]
conducting yet another experiment on my Gigabyte board, I remove the UAD driver along with the HDX_GigabyteExt_Creative_RTK.inf file, reboot and reinstalled the UAD driver but this time using a modded **HDX_GigabyteExt_DOLBY_RTK.inf** file by adding my ID "1458A182" only on there [no addition of the "OMFX" entries]

result - Dolby Atmos app and Realtek Audio Console apps running but this time the Realtek equalizer & environment options no longer available (similar to the Lenovo, Realtek UAD + Dolby problem that wkonowrocki is having)

![realtek-uwp-dolbyatmos-uwp-soundoptions](https://user-images.githubusercontent.com/61091896/76592079-74366e00-64af-11ea-9aa9-845c5e58e22f.png)

**Comment by pal1000** [comment]
I already extended patching functionality. Can any of you post those OMFX entries in INF format?

**Comment by SesioN** [comment]
Also missing Realtek EQ.

MSI Mainboard with ALC1220P-VB2 Codec:
HDAUDIOFUNC_01&VEN_10EC&DEV_1168&SUBSYS_1462CB17&REV_1001

**Comment by shibajee** [comment]
@epman100 Try this https://gist.github.com/shibajee/ada8f23acfe6260671378aee4a40380e
both realtek default effect and dolby should work.

**Comment by Hazbelll** [comment]
> I already extended patching functionality. Can any of you post those OMFX entries in INF format?

That was quick, I figured there was more to do as you said it would take a while! Not too sure what you mean by OMFX Entries, but I can try to find mine if given directions?

**Comment by epman100** [comment]
> 
> 
> I already extended patching functionality. Can any of you post those OMFX entries in INF format?

actually there is no need for me to post them as I checked the contents of the **HDX_GenericExt_RTK.inf** file and the OMFX entries are **already there**.  it's just that those specific entries have been removed in certain OEM based extension INF files like HDX_GigabyteExt_DOLBY_RTK.inf, HDX_ClevoExt_CT_RTK.inf, HDX_LenovoExt_DOLBY_FORTE.inf etc.

if you don't believe me compare the contents of the "HDX_GenericExt_RTK.inf" file with another OEM based extension INF like "HDX_LenovoExt_DOLBY_FORTE.inf" using a file comparison tool like WinDiff.exe.  I use Windiff.exe v6.1.7716.0 (google search for that) to compare two different extension INF files - Windiff will highlight the differences.

**Comment by SesioN** [comment]
@epman100 @pal1000 

https://us.msi.com/Motherboard/MPG-Z390-GAMING-PRO-CARBON/Specification

**Comment by eebssk1** [comment]
The equalizer and some other stuff  in realtek control panel require hardware dsp available in PCH and soldered to the codec(thus called 'offload').
Only a few devices support it nowadays.You should use the sound effect software provided by OEM or go for EqualizerAPO.

**Comment by eebssk1** [comment]
> The equalizer and some other stuff in realtek control panel require hardware dsp available in PCH and soldered to the codec(thus called 'offload'). Only a few devices support it nowadays.You should use the sound effect software provided by OEM or go for EqualizerAPO.

It  looks like software mode is still supported.
You can force enable the EQ with following steps(this will eat some CPU and memory without hardware support).
![图片](https://user-images.githubusercontent.com/16506472/192128653-888cfee5-9ecd-4395-8535-26c9efb80028.png)
![图片](https://user-images.githubusercontent.com/16506472/192128664-286c095c-a4b8-4f2e-8a39-8d8158f7a238.png)
![图片](https://user-images.githubusercontent.com/16506472/192128672-b8dc6f65-de69-43f0-99a7-3f1a68ba29cf.png)

@pal1000 
@SesioN

**Comment by cemck** [comment]
@eebssk1 where can I get the Test Utility?

**Comment by eebssk1** [comment]
> @eebssk1 where can I get the Test Utility?

I can't find the original url, But I uploaded it to my web server.
https://sh2.godaftwithebk.pub/nextcloud/s/LMCy6DLGARaiHj5

**Comment by cemck** [comment]
Thanks! @eebssk1 i have this UI since the latest driver version 6.0.9244.1 and ui version 1.29.256.0:
![Screenshot 2022-11-27 221724](https://user-images.githubusercontent.com/35745629/204160188-6ae815c5-dfb1-403f-8df1-fca91918d94f.png)
It was possible to get the Equalizer to show up after uninstalling the DTS App which is downloading after driver installation, but with the latest it is always showing the DTS UI.
Do you know how to get the equalizer UI to show up again with test utility tweaking? The steps you provided doesnt seem to work for me

**Comment by eebssk1** [comment]
> Thanks! @eebssk1 i have this UI since the latest driver version 6.0.9244.1 and ui version 1.29.256.0: ![Screenshot 2022-11-27 221724](https://user-images.githubusercontent.com/35745629/204160188-6ae815c5-dfb1-403f-8df1-fca91918d94f.png) It was possible to get the Equalizer to show up after uninstalling the DTS App which is downloaded after driver installation, but now with the latest driver it is always showing the DTS UI. Do you know how to get the equalizer UI to show up again with test utility tweaking? The steps you provided doesnt seem to work for me
> 
> Older driver versions are not working anymore with Windows 11

No, it's only for generic driver or devices without 3rd effects.

**Comment by cemck** [comment]
> > Thanks! @eebssk1 i have this UI since the latest driver version 6.0.9244.1 and ui version 1.29.256.0: ![Screenshot 2022-11-27 221724](https://user-images.githubusercontent.com/35745629/204160188-6ae815c5-dfb1-403f-8df1-fca91918d94f.png) It was possible to get the Equalizer to show up after uninstalling the DTS App which is downloaded after driver installation, but now with the latest driver it is always showing the DTS UI. Do you know how to get the equalizer UI to show up again with test utility tweaking? The steps you provided doesnt seem to work for me
> > Older driver versions are not working anymore with Windows 11
> 
> No, it's only for generic driver or devices without 3rd effects.

Okay I did switch to the generic one, know its working out of box. Is there any hidden tweaking option to get an equalizer with more detailed frequency range?

**Comment by eebssk1** [comment]
> > > Thanks! @eebssk1 i have this UI since the latest driver version 6.0.9244.1 and ui version 1.29.256.0: ![Screenshot 2022-11-27 221724](https://user-images.githubusercontent.com/35745629/204160188-6ae815c5-dfb1-403f-8df1-fca91918d94f.png) It was possible to get the Equalizer to show up after uninstalling the DTS App which is downloaded after driver installation, but now with the latest driver it is always showing the DTS UI. Do you know how to get the equalizer UI to show up again with test utility tweaking? The steps you provided doesnt seem to work for me
> > > Older driver versions are not working anymore with Windows 11
> > 
> > No, it's only for generic driver or devices without 3rd effects.
> 
> Okay I did switch to the generic one, now its working out of the box. Is there any hidden tweaking option to get an equalizer with more detailed frequency range?

In my experience, No, only lesser.

**Comment by eebssk1** [comment]
> > > > Thanks! @eebssk1 i have this UI since the latest driver version 6.0.9244.1 and ui version 1.29.256.0: ![Screenshot 2022-11-27 221724](https://user-images.githubusercontent.com/35745629/204160188-6ae815c5-dfb1-403f-8df1-fca91918d94f.png) It was possible to get the Equalizer to show up after uninstalling the DTS App which is downloaded after driver installation, but now with the latest driver it is always showing the DTS UI. Do you know how to get the equalizer UI to show up again with test utility tweaking? The steps you provided doesnt seem to work for me
> > > > Older driver versions are not working anymore with Windows 11
> > > 
> > > No, it's only for generic driver or devices without 3rd effects.
> > 
> > Okay I did switch to the generic one, now its working out of the box. Is there any hidden tweaking option to get an equalizer with more detailed frequency range?
> 
> In my experience, No, only lesser.

As I told, if you need more bands or need other usage, you should use equalizer apo instead.It eats CPU but has nice features.
